cpp/src/strings/convert/convert_fixed_point.cu

```diff
@@ -33,74 +33,152 @@
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
 
+#include <thrust/optional.h>
 #include <thrust/transform.h>
 
 namespace cudf {
 namespace strings {
 namespace detail {
 namespace {
+
+struct string_to_decimal_base {
+  /**
+   * @brief Return the integer component of a decimal string.
+   *
+   * This is reads everything up to the exponent 'e' notation.
+   * The return includes the integer digits and any exponent offset.
+   *
+   * @param[in,out] iter Start of characters to parse
+   * @param[in] end End of characters to parse
+   * @return Integer component and exponent offset.
+   */
+  __device__ thrust::pair<uint64_t, int> parse_integer(char const*& iter,
+                                                       char const* iter_end) const
+  {
+    // highest value where another decimal digit cannot be appended without an overflow;
+    // this preserves the most digits when scaling the final result
+    constexpr uint64_t decimal_max = (std::numeric_limits<uint64_t>::max() - 9L) / 10L;
+
+    uint64_t value     = 0;  // for checking overflow
+    int exp_offset     = 0;
+    bool decimal_found = false;
+
+    while (iter < iter_end) {
+      auto const ch = *iter++;
+      if (ch == '.' && !decimal_found) {
+        decimal_found = true;
+        continue;
+      }
+      if (ch < '0' || ch > '9') {
+        --iter;
+        break;
+      }
+      if (value > decimal_max) {
+        exp_offset += static_cast<int>(!decimal_found);
+      } else {
+        value = (value * 10) + static_cast<uint64_t>(ch - '0');
+        exp_offset -= static_cast<int>(decimal_found);
+      }
+    }
+    return {value, exp_offset};
+  }
+
+  /**
+   * @brief Return the exponent of a decimal string.
+   *
+   * This should only be called after the exponent 'e' notation was detected.
+   * The return is the exponent (base-10) integer and can only be
+   * invalid if `check_only == true` and invalid characters are found or the
+   * exponent overflows an int32.
+   *
+   * @tparam check_only Set to true to verify the characters are valid and the
+   *         exponent value in the decimal string does not overflow int32
+   * @param[in,out] iter Start of characters to parse
+   *                     (points to the character after the 'E' or 'e')
+   * @param[in] end End of characters to parse
+   * @return Integer value of the exponent
+   */
+  template <bool check_only = false>
+  __device__ thrust::optional<int> parse_exponent(char const* iter, char const* iter_end) const
```

Comment:
Should you call out that this int is 32-bit? I know this is pedantic at best though.

Reply:
Yes, I think you are right. I should change these to `int32_t` to make it more clear.